physicsnemo/models/layers/ball_query.py

```diff
@@ -26,244 +237,91 @@ class BallQuery(torch.autograd.Function):
-        wp.launch(
-            BallQuery.sparse_ball_query,
-            inputs=[
-                ctx.points2,
-                ctx.mapping,
-                ctx.num_neighbors,
-            ],
-            outputs=[
-                ctx.outputs,
-            ],
-            dim=ctx.dim,
+        # Apply the primitive.  Note the batch index is removed.
+        mapping, num_neighbors, outputs = _ball_query_forward_primative_(
+            points1[0],
+            points2[0],
+            k,
+            radius,
+            hash_grid,
         )
+        ctx.save_for_backward(points1, points2, mapping, num_neighbors, outputs)
 
-        return (
-            wp.to_torch(ctx.mapping),
-            wp.to_torch(ctx.num_neighbors),
-            wp.to_torch(ctx.outputs),
-        )
+        return mapping, num_neighbors, outputs
 
     @staticmethod
-    def backward(
-        ctx,
-        grad_mapping: torch.Tensor,
-        grad_num_neighbors: torch.Tensor,
-        grad_outputs: torch.Tensor,
-    ) -> tuple[torch.Tensor, torch.Tensor, None, None, None]:
-        # Map incoming torch grads to our output variable
-        ctx.outputs.grad = wp.from_torch(grad_outputs, dtype=wp.float32)
-
-        # Run the kernel in adjoint mode
-        wp.launch(
-            BallQuery.sparse_ball_query,
-            inputs=[
-                ctx.points2,
-                ctx.mapping,
-                ctx.num_neighbors,
-            ],
-            outputs=[
-                ctx.outputs,
-            ],
-            adj_inputs=[ctx.points2.grad, ctx.mapping.grad, ctx.num_neighbors.grad],
-            adj_outputs=[
-                ctx.outputs.grad,
-            ],
-            dim=ctx.dim,
-            adjoint=True,
+    def backward(ctx, grad_mapping, grad_num_neighbors, grad_outputs):
+
+        points1, points2, mapping, num_neighbors, outputs = ctx.saved_tensors
+        # Apply the primitive
+        p2_grad = _ball_query_backward_primative_(
+            points1[0],
+            points2[0],
+            mapping,
+            num_neighbors,
+            outputs,
+            grad_mapping,
+            grad_num_neighbors,
+            grad_outputs,
         )
+        p2_grad = p2_grad.unsqueeze(0)
 
         # Return the gradients
         return (
-            wp.to_torch(ctx.points1.grad).unsqueeze(0),
-            wp.to_torch(ctx.points2.grad).unsqueeze(0),
-            # None,
-            # None,
+            torch.zeros_like(points1),
+            p2_grad,
             None,
             None,
             None,
         )
 
 
+def ball_query_layer(
+    points1: torch.Tensor,
+    points2: torch.Tensor,
+    k: int,
+    radius: float,
+    hash_grid: wp.HashGrid,
+):
+    """
+    Wrapper for BallQuery.apply to support a functional interface.
+    """
+    if has_torch_function((points1, points2)):
+        return handle_torch_function(
+            ball_query_layer, (points1, points2), points1, points2, k, radius, hash_grid
+        )
```

Comment:
This is needed to enable torch_function to work on this operation, and therefore overload in ShardTensor.